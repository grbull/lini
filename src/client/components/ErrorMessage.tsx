import cn from 'classnames';
import React, { ReactElement, ReactNode } from 'react';

interface Props {
  children: string | ReactNode;
  className?: string;
  illustration?: boolean;
}

export function ErrorMessage({
  children,
  className,
  illustration,
}: Props): ReactElement {
  return (
    <div
      className={cn(
        'px-2',
        'py-2.5',
        'bg-red-400',
        'bg-opacity-50',
        'border',
        'border-solid',
        'border-red-400',
        className
      )}
    >
      {illustration && (
        <div
          className={cn('flex', 'flex-col', 'items-center', 'justify-center')}
        >
          <img
            alt="Frustrated Illustration"
            className={cn('opacity-80', 'mb-2.5')}
            src="frustrated.png"
            width={80}
          />
        </div>
      )}

      <h2 className={cn('font-bold')}>Error</h2>
      <p>{children}</p>
    </div>
  );
}

ErrorMessage.defaultProps = { className: undefined, illustration: false };
