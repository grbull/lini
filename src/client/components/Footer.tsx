import cn from 'classnames';
import React, { ReactElement } from 'react';

import packageJson from '../../../package.json';

export function Footer(): ReactElement {
  return (
    <div
      className={cn(
        'h-8',
        'fixed',
        'bottom-0',
        'hidden md:block',
        'w-full',
        'bg-gray-100 dark:bg-gray-600'
      )}
    >
      <div
        className={cn(
          'max-w-4xl',
          'h-full',
          'mx-auto',
          'flex',
          'items-center',
          'justify-center',
          'font-light',
          'text-sm'
        )}
      >
        <span>Lini 2021 &copy;</span>
        <span className={cn('px-4')}>•</span>
        <span>Version: {packageJson.version}</span>
        <span className={cn('px-4')}>•</span>
        <a className={cn('underline')} href="https://github.com/grbull/lini">
          Source
        </a>
      </div>
    </div>
  );
}
