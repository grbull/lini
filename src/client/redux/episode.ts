/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
