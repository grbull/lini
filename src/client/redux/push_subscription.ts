/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { api } from '../utils/api';

type PushSubscriptionState = {
  status: 'loading' | 'idle' | 'error';
  error?: string;
};

const initialState: PushSubscriptionState = {
  status: 'idle',
};

export const create = createAsyncThunk<void, PushSubscriptionCreateDto>(
  'push_subscription/get',
  (createDto) => api.push.create(createDto)
);

export const pushSubscription = createSlice({
  name: 'push_subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state) => {
      state.status = 'idle';
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const pushSubscriptionActions = {
  ...pushSubscription.actions,
  create,
};
