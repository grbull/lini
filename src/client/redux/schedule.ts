/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type ScheduleState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: {
    future: EpisodeDto[];
    past: EpisodeDto[];
  };
  error?: string;
};

const initialState: ScheduleState = {
  status: 'init',
  data: {
    future: [],
    past: [],
  },
};

export const get = createAsyncThunk<EpisodeScheduleDto>('schedule/get', () =>
  api.episode.getSchedule()
);

export const schedule = createSlice({
  name: 'schedule',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const scheduleActions = {
  ...schedule.actions,
  get,
};
