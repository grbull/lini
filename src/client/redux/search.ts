/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type SearchState = {
  isLoading: boolean;
  error?: string;
  data?: ShowDto[];
};

const initialState: SearchState = {
  isLoading: false,
};

export const query = createAsyncThunk<ShowDto[], string>(
  'search/query',
  (query) => api.show.search(query)
);

export const search = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(query.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(query.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(query.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const searchActions = { ...search.actions, query };
