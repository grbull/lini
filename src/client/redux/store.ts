import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
