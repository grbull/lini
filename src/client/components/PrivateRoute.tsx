import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { Redirect, Route, RouteProps } from 'react-router-dom';

import { RootState } from '../redux/store';

export function PrivateRoute({ ...rest }: RouteProps): ReactElement {
  const user = useSelector((state: RootState) => state.user);

  return user.data && !user.error ? (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <Route {...rest} />
  ) : (
    <Redirect to="/login" />
  );
}
