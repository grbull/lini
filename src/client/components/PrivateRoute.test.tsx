/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';
import { ErrorMessage } from './ErrorMessage';
import { Footer } from './Footer';
import { PrivateRoute } from './PrivateRoute';

function setup(state?: Partial<RootState>): RenderResult {
  const history = createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: state ?? undefined,
  });

  const utils = render(
    <Router history={history}>
      <Provider store={store}>
        <Route component={ErrorMessage} exact path="/login" />
        <PrivateRoute component={Footer} />
      </Provider>
    </Router>
  );

  return { ...utils };
}

describe('PrivateRoute Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup({
      user: {
        status: 'idle',
        data: {
          theme: 'auto',
          email: 'test@email.com',
          notifications: false,
          dateCreated: '',
          dateModified: '',
        },
      },
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot with user error', () => {
    const { asFragment } = setup({
      user: { status: 'error', error: 'Generic Error' },
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
