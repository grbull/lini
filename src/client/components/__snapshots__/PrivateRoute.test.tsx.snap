// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-8 fixed bottom-0 hidden md:block w-full bg-gray-100 dark:bg-gray-600"
  >
    <div
      class="max-w-4xl h-full mx-auto flex items-center justify-center font-light text-sm"
    >
      <span>
        Lini 2021 ©
      </span>
      <span
        class="px-4"
      >
        •
      </span>
      <span>
        Version: 2.11.0
      </span>
      <span
        class="px-4"
      >
        •
      </span>
      <a
        class="underline"
        href="https://github.com/grbull/lini"
      >
        Source
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PrivateRoute Component matches the snapshot with user error 1`] = `
<DocumentFragment>
  <div
    class="px-2 py-2.5 bg-red-400 bg-opacity-50 border border-solid border-red-400"
  >
    <h2
      class="font-bold"
    >
      Error
    </h2>
    <p />
  </div>
</DocumentFragment>
`;
